### PR TITLE
Add user routes , including:

### PR DESCRIPTION
**Pull request by Melody-Le**

### CODE REVIEW

File: controllers/userController.js [hunk 1]
```diff
@@ -64,10 +130,30 @@ const controller = {
...
         .lean();
-      const appliedJobs = await ContributorRelationshipModel.find(
+      const userProjectApplied = [];
+      for (let i = 0, len = appliedJobs.length; i < len; i++) {
+        userProjectApplied[i] = appliedJobs[i]?.contributor_id?.project_id;
+      }
+
+      return res.json(userProjectApplied);
+    } catch (error) {
+      console.log(error);
```
Comment (ZoeNguyen73) [inline controllers/userController.js]: can remove console.log here

File: controllers/userController.js [hunk 2]
```diff
@@ -77,40 +163,55 @@ const controller = {
...
+  },
+  showUserProjectsFollowing: async (req, res) => {
+    const username = req.params.username;
+    try {
+      const profileOwner = await UserModel.findOne({ username }, { __v: 0, hash: 0 }).lean();
+      if (!profileOwner) {
+        return res.status(404).json();
       }
       const followingProjectsRelationship = await ProjectsRelationshipModel.find(
         { user_id: profileOwner._id, state: 'published' },
```
Comment (ZoeNguyen73) [inline controllers/userController.js]: there is no state under ProjectsRelationshipModel right?

File: controllers/userController.js [hunk 3]
```diff
@@ -77,40 +163,55 @@ const controller = {
...
+        error: 'Failed to fetch contributed projects of this user from database',
       });
     }
   },
+
   editProfile: async (req, res) => {
+    const profileOwner = await UserModel.findOne({ username: req.authUser.username });
+    const user = await UserModel.findOne({ username: req.params.username });
+    // authorisation check: whether user is the project owner
+    if (user?._id.toString() !== profileOwner._id.toString()) {
```
Comment (ZoeNguyen73) [inline controllers/userController.js:204]: should put your authorization code inside userAuth middleware instead

File: controllers/userController.js [hunk 4]
```diff
@@ -150,23 +255,16 @@ const controller = {
...
-        return res.status(401).json({
-          error: 'User is not authorised to change this profile',
-        });
-      }
-      const skillsArr = skills
-        ?.split(',')
-        .map((item) => item.trim())
-        .filter((item) => validSkills.includes(item));
-      const interestsArr = interests?.split(',').map((item) => item.trim());
+      // const skillsArr = skills
```
Comment (ZoeNguyen73) [inline controllers/userController.js]: can remove this code since commented out

File: routes/userRoutes.js [hunk 5]
```diff
@@ -10,14 +10,22 @@ router.post('/:token/activate', userController.activateAccount);
 router.get('/:username/following', userController.showFollowingUsers);
 router.get('/:username/followers', userController.showFollowerUsers);
 router.post('/:username/follow', userAuth.isAuthenticated, userController.followUser);
+
+router.get('/:username', userController.showUserProfile);
+router.get('/:username/projects', userController.showUserProjects);
+router.get('/:username/projects/public', userController.showUserProjectsPublic);
+router.get('/:username/projects/draft', userController.showUserProjectsDraft);
```
Comment (ZoeNguyen73) [inline routes/userRoutes.js:17]: for drafts, applied >> these are private and will only be displayed to the user themselves right?
if that is the case, will need authorization here

File: controllers/userController.js (hunk 1 above)
Comment (Melody-Le) [reply controllers/userController.js]: updated

File: controllers/userController.js (hunk 2 above)
Comment (Melody-Le) [reply controllers/userController.js]: updated

File: controllers/userController.js (hunk 4 above)
Comment (Melody-Le) [reply controllers/userController.js]: updated

File: routes/userRoutes.js (hunk 5 above)
Comment (Melody-Le) [reply routes/userRoutes.js:17]: this one, can you help me put when you help me do the userIsAuthorised?

**Review by ZoeNguyen73 — APPROVED** [review]

**Pull request merged**